### ISSUE TITLE
Generate Kotlin and Swift at the same time as generating TS/C++

### ISSUE THREAD

**Issue by jhugman** [jhugman/uniffi-bindgen-react-native#149]
@zzorba has been asking for this for a lilttle while.

While the main React Native app requires typescript and C++, iOS extensions still require Swift. 

Can we generate the Swift from the Rust at the same time as generating the Typescript and C++?

This issue just covers the generating and building of the Kotlin and Swift generation.

@almost adds this question: can we run the Kotlin and Swift bindings at the same time as the Typescript? This enables a host of new capabilities.

**Comment by jhugman** [comment]
> I believe you can run the bindings at the same time, though I do not know if they share the process state or not (I would assume they would, but I'm just not sure since it hasn't been our need).

Interesting.

Eyeballing the code, I was under the impression that just generating the extra bindings was necessary but not sufficient for running in the same process. I am expecting that this would require changes to how [callback interfaces](https://github.com/mozilla/uniffi-rs/blob/main/uniffi_macros/src/export/callback_interface.rs) work.

Still, that is for another issue :)

**Comment by Johennes** [comment]
Don't app extensions on iOS run in their own process (and with limited system resources) anyway? From what I know they can only share certain storage locations with the main app but not the same system process.

**Comment by zzorba** [comment]
Yes that's correct about extensions, they are isolated and only share storage and keychain access.

But theoretically you could have swift code and react native code in the library both talking to the same binary.

Not sure if it's a large use case though?

**Comment by Johennes** [comment]
Oh, I see what you mean now. Hm, yeah I suppose there could be a use case when you want to combine the bindings with some native APIs without having to poke out to TS and back in to another library.

**Comment by Johennes** [comment]
@jhugman I'm curious about your thoughts on how to best integrate this into ubrn, starting with the Swift side. We could just call `generate_bindings` with `uniffi_bindgen::bindings::SwiftBindingGenerator` in [`cli.rs`](https://github.com/jhugman/uniffi-bindgen-react-native/blob/main/crates/ubrn_bindgen/src/cli.rs#L148). The Swift bindings generator produces headers and module maps that need to go into the `.xcframework`, however, and that has already been assembled at this point as part of the `build ios` command.

The build command currently has a [`no_xcodebuild`](https://github.com/jhugman/uniffi-bindgen-react-native/blob/main/crates/ubrn_cli/src/ios.rs#L126) to allow constructing the framework externally. I'm wondering if we could lift this and instead move the framework assembly into the `generate` command?

Alternatively we could also reassemble the framework during `generate` or just push the files around ourselves (the framework is just a folder). That might be a little backwards though.

**Comment by Johennes** [comment]
> If we need to generate Swift before the xframework is made— are we going to have to invoke a swift compiler?

No, I don't think so. It's basically all done by `SwiftBindingGenerator`. It takes a library file from within the xcframework as its input. The Swift itself lives outside the framework. However, the bindings generator also produces some headers and these need to go _into_ the framework for the Swift to work. That's what's causing the ordering hassle.

> I think my preference would be to generate the swift as part of the build, rather than moving the lipo/xcodebuild part to the generate step.

Ok, let me try it like this because I think it'll also be less changes than the other option. If needed, we can always consider rearranging the steps later.

**Comment by Johennes** [comment]
This has been addressed now with #214 and #218.
